extensions/google-java/src/main/kotlin/com/connectrpc/extensions/GoogleJavaProtoAdapter.kt

```diff
@@ -17,56 +17,49 @@ package com.connectrpc.extensions
 import com.connectrpc.CODEC_NAME_PROTO
 import com.connectrpc.Codec
 import com.google.protobuf.CodedOutputStream
-import com.google.protobuf.GeneratedMessageV3
+import com.google.protobuf.ExtensionRegistryLite
 import com.google.protobuf.Internal
+import com.google.protobuf.MessageLite
 import okio.Buffer
 import okio.BufferedSource
-import java.io.IOException
 import kotlin.reflect.KClass
+import kotlin.reflect.cast
 
 /**
  * Adapter to use Google's protobuf-java runtime for
  * deserializing and serializing data types.
  */
-internal class GoogleJavaProtoAdapter<E : GeneratedMessageV3>(
-    clazz: KClass<E>,
+internal class GoogleJavaProtoAdapter<E : MessageLite>(
+    private val clazz: KClass<E>,
+    private val registry: ExtensionRegistryLite,
 ) : Codec<E> {
-    /**
-     * Casting assumes the user is using Google's GeneratedMessageV3 type.
-     */
-    @Suppress("UNCHECKED_CAST")
-    private val instance by lazy {
-        Internal.getDefaultInstance(clazz.java as Class<GeneratedMessageV3>)
-    }
+    private val instance: E
+        get() = Internal.getDefaultInstance(clazz.java)
 
     override fun encodingName(): String {
         return CODEC_NAME_PROTO
     }
 
-    /**
-     * Casting assumes the user is using Google's GeneratedMessageV3 type.
-     * The builder returns a GeneratedMessageV3 but it is assumed to be used
-     * with the assumption that the generic E is the underlying type.
-     */
-    @Suppress("UNCHECKED_CAST")
     override fun deserialize(source: BufferedSource): E {
-        return instance.parserForType.parseFrom(source.inputStream()) as E
+        return clazz.cast(instance.parserForType.parseFrom(source.inputStream(), registry))
     }
 
     override fun serialize(message: E): Buffer {
-        return Buffer().write(message.toByteArray())
+        return serialize(message, false)
     }
 
     override fun deterministicSerialize(message: E): Buffer {
-        return try {
-            val result = Buffer()
-            val output = CodedOutputStream.newInstance(result.outputStream())
+        return serialize(message, true)
+    }
+
+    private fun serialize(message: E, deterministic: Boolean): Buffer {
```

Comment:
I made both deterministic and non-deterministic flows go through the same flow to make sure they both worked. Previously, the deterministic flow had several serious problems that meant that Connect GET didn't actually work at all:
1) It used `Buffer.outputStream()` to write the marshalled data. However, that method returns the buffer's _underlying buffered sink_, or a no-op stream if there isn't one. We create a new buffer with default constructor, so it has no sink. So the resulting buffers were _always_ empty 😭 
2) It called `output.checkNoSpaceLeft`, but the coded stream was created using an output stream, not a byte array, so this method _always threw an exception_ ([link](https://github.com/protocolbuffers/protobuf/blob/main/java/core/src/main/java/com/google/protobuf/CodedOutputStream.java#L2216-L2218)).

I've left the call to `checkNoSpaceLeft` since now we're creating the output stream with a byte array (which was necessary since `Buffer.outputStream()` doesn't actually do the right thing).

One possible way to improve this and remove this allocation/copy of bytes would be to write an adapter around an `okio.Buffer` that implements [`ByteOutput`](https://github.com/protocolbuffers/protobuf/blob/main/java/core/src/main/java/com/google/protobuf/ByteOutput.java).